memcache/pool.go

```diff
@@ -0,0 +1,294 @@
+// It also periodically closes idle connections to minimize unused resources.
+// Note: Any connection obtained from the pool must be returned to the pool when done (either using the Put or Release method).
+type Pool struct {
+	addr   net.Addr
+	dialer func(ctx context.Context, network, addr string) (net.Conn, error)
+	config PoolConfig
+
+	mu     sync.Mutex
+	idle   []idleConn
+	active chan struct{}
+	hits   atomic.Int64
+	misses atomic.Int64
+}
+
+type idleConn struct {
+	c *Conn
+	t time.Time // last used time
+}
+
+func NewPool(addr net.Addr, dial dialer, config PoolConfig) *Pool {
+	// ensure config has defaults
+	config.withDefaults()
+
+	p := &Pool{
+		addr:   addr,
+		dialer: dial,
+		config: config,
+	}
+
+	// start a goroutine to close idle connections periodically
+	go func() {
+		ticker := time.NewTicker(p.config.IdleClosePeriod)
+		defer ticker.Stop()
+		for range ticker.C {
+			p.closeIdle()
+		}
+	}()
+
+	p.active = make(chan struct{}, p.config.MaxConns)
+	return p
+}
+
+// Get returns a connection from the pool or creates a new one if necessary.
+// The connection must be returned to the pool when done.
+// If max connections are already in use, Get will block until a connection is available or the context is canceled.
+func (p *Pool) Get(ctx context.Context) (*Conn, error) {
+	// first acquire a slot
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case p.active <- struct{}{}:
+	}
+
+	// then try to get an idle connection
+	if c := p.getIdle(); c != nil {
+		p.hits.Add(1)
+		return c, nil
+	}
+	p.misses.Add(1)
+
+	// if no idle connection, create a new one
+	nc, err := p.dialer(ctx, p.addr.Network(), p.addr.String())
+	if err != nil {
+		// release the slot
+		<-p.active
+
+		var ne net.Error
+		if errors.As(err, &ne) && ne.Timeout() {
+			return nil, &ConnectTimeoutError{p.addr}
+		}
+		return nil, err
+	}
+
+	return &Conn{
+		nc:   nc,
+		addr: p.addr,
+		rw:   bufio.NewReadWriter(bufio.NewReader(nc), bufio.NewWriter(nc)),
+		pool: p,
+	}, nil
+}
+
+func (p *Pool) getIdle() *Conn {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	if len(p.idle) == 0 {
+		return nil
+	}
+
+	// idle is a stack (LIFO)
+	ic := p.idle[len(p.idle)-1]
+	p.idle = p.idle[:len(p.idle)-1]
+	return ic.c
+}
+
+// Put returns a connection to the pool for reuse.
+func (p *Pool) Put(c *Conn) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	// idle is a stack (LIFO)
+	p.idle = append(p.idle, idleConn{c: c, t: time.Now()})
+	<-p.active
+}
+
+// Release releases a slot in the pool. This should be called if a connection was closed outside the pool
+// (most likely due to an error).
+func (p *Pool) Release() {
+	<-p.active
+}
+
+// closeIdle closes idle connections that have been idle for longer than the max idle lifetime.
+func (p *Pool) closeIdle() {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	idle := make([]idleConn, 0, len(p.idle))
+	for _, ic := range p.idle {
+		if time.Since(ic.t) > p.config.MaxIdleLifetime {
+			ic.c.nc.Close()
```

Comment:
should we consider any error handling on the closes?

Reply:
what would you do?

Reply:
🤔 -- I usually think of it in one of three ways:
- log an error. This can suck b/c the close is so separate from the context of it
- have an error channel. This usually ends up looking like ^^ with more steps 🙃 
- have an error metric

Reply:
I don't consider this a blocker, but something to think on.